x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/SingleFieldFullTextFunction.java

```diff
@@ -0,0 +1,257 @@
+        return Expressions.isGuaranteedNull(field());
+    }
+
+    @Override
+    public Object fold(FoldContext ctx) {
+        // We only fold when the field is null (it's not present in the mapping), so we return null
+        return Literal.NULL;
+    }
+
+    @Override
+    public Nullability nullable() {
+        // The function is nullable if the field is guaranteed to be null, due to the field not being present in the mapping
+        return Expressions.isGuaranteedNull(field()) ? Nullability.TRUE : Nullability.FALSE;
+    }
+
+    @Override
+    public BiConsumer<LogicalPlan, Failures> postAnalysisPlanVerification() {
+        return (plan, failures) -> {
+            super.postAnalysisPlanVerification().accept(plan, failures);
+            fieldVerifier(plan, this, field, failures);
+        };
+    }
+
+    @Override
+    public BiConsumer<LogicalPlan, Failures> postOptimizationPlanVerification() {
+        // check plan again after predicates are pushed down into subqueries
+        return (plan, failures) -> {
+            super.postOptimizationPlanVerification().accept(plan, failures);
+            fieldVerifier(plan, this, field, failures);
+        };
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        // Functions do not serialize options, as they get included in the query builder.
+        // We override equals and hashcode to ignore options when comparing two function instances
+        if (o == null || getClass() != o.getClass()) return false;
+        SingleFieldFullTextFunction that = (SingleFieldFullTextFunction) o;
+        return Objects.equals(field(), that.field())
+            && Objects.equals(query(), that.query())
+            && Objects.equals(queryBuilder(), that.queryBuilder());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(field(), query(), queryBuilder());
+    }
+
+    /**
+     * Returns the set of allowed data types for the field parameter.
+     * Each subclass defines which field types it supports.
+     */
+    protected abstract Set<DataType> getFieldDataTypes();
+
+    /**
+     * Returns the set of allowed data types for the query parameter.
+     * Each subclass defines which query types it supports.
+     */
+    protected abstract Set<DataType> getQueryDataTypes();
+
+    /**
+     * Returns the allowed options map for this function.
+     * Keys are option names, values are the expected data types.
+     */
+    protected abstract Map<String, DataType> getAllowedOptions();
+
+    /**
+     * Returns a human-readable string listing the expected field types.
+     * Used in error messages.
+     */
+    protected String getExpectedFieldTypesString() {
+        return expectedTypesAsString(getFieldDataTypes());
+    }
+
+    /**
+     * Returns a human-readable string listing the expected query types.
+     * Used in error messages.
+     */
+    protected String getExpectedQueryTypesString() {
```

Comment:
```suggestion
    protected String expectedQueryTypesString() {
```

Reply:
Done as well in ad811e5cb8994eade1d5ca017468136d7785b06d